README.md
# Knife ESX

## DESCRIPTION:

This is the unofficial Opscode Knife plugin for ESX. This plugin gives knife the ability to create, bootstrap, and manage virtual machines in a VMWare ESX/ESXi host.

You'll need an ESX(i)4/5 host with SSH enabled and a privileged user account to connect to it via SSH.

You'll also need a base VM template (a VMDK in fact) with CentOS/Ubuntu/Fedora and vmware-tools installed.

If you're using VMWare ESXi with a free license, don't forget the `--free-license` option or you'll get the following error:

    ERROR: RbVmomi::Fault: RestrictedVersion: Current license or ESXi version prohibits execution of the requested operation

Home page: http://github.com/maintux/knife-esx

## INSTALLATION:

Be sure you are running the latest version Chef. Versions earlier than 0.10.0 don't support plugins:

    gem install chef

This plugin is distributed as a Ruby Gem. To install it, run:

    gem install knife-esx

Depending on your system's configuration, you may need to run this command with root privileges.

## CONFIGURATION:

In order to communicate with the ESX Cloud API you will have to tell Knife about your Username and API Key.  The easiest way to accomplish this is to create some entries in your `knife.rb` file:

    knife[:esx_host] = "Your ESX host address"
    knife[:esx_username]  = "root"
    knife[:esx_password]  = "secret"

You also have the option of passing your ESX Host/Username/Password into the individual knife subcommands using the `--esx-host` `--esx-username` and `--esx-password` command options

## COMMAND LINE BASIC OPTIONS

<pre style="word-wrap: normal;">
--esx-host HOST                           The ESX host to connect to (default: 127.0.0.1)
--esx-username USERNAME                   The ESX username used to connect to the host (default: root)
--esx-password PASSWORD                   The ESX user password
--esx-templates-dir TEMPLATES_DIRECTORY   The folder where the VM template is stored
--free-license                            If the ESX(i) host has a free license (default: false)
--insecure                                Use an insecure connection (default: true)
</pre>

For subcommand specific options, refer to the related subcommand

## SUBCOMMANDS:

This plugin provides the following Knife subcommands.  Specific command options can be found by invoking the subcommand with a `--help` flag

### knife esx vm create

Provisions a new virtual machine in the ESX host and then perform a Chef bootstrap (using the SSH protocol).  The goal of the bootstrap is to get Chef installed on the target system so it can run Chef Client with a Chef Server. The main assumption is a baseline OS installation exists (provided by the provisioning). It is primarily intended for Chef Client systems that talk to a Chef server.  By default the virtual machine is bootstrapped using the {ubuntu10.04-gems}[https://github.com/opscode/chef/blob/master/chef/lib/chef/knife/bootstrap/ubuntu10.04-gems.erb] template.  This can be overridden using the `-d` or `--template-file` command options.

#### _Command line options_

<pre style="word-wrap: normal;">
--vm-disk FILE                                            The path to the VMDK disk file
--vm-name NAME                                            The Virtual Machine name
--vm-cpus CPUS                                            The Virtual Machine total number of virtual cpus (for the calculation see below) (default: 1)
--vm-cpu-cores CPU_CORES                                  The number of cores per CPU socket. The number of sockets is calculated as &lt;CPUS&gt;=&lt;SOCKETS&gt;*&lt;CPU_CORES&gt; (default: 1)
--datastore NAME                                          The Datastore to use for the VM files (default: datastore1)
--guest-id NAME                                           The VM GuestID (default: otherGuest)
--vm-memory MEM                                           The VM memory in MB (default: 512)
-N NAME | --node-name NAME                                The Chef node name for your new node
--prerelease                                              Install the pre-release chef gems
--bootstrap-version VERSION                               The version of Chef to install
-d DISTRO | --distro DISTRO                               Bootstrap a distro using a template (default: ubuntu10.04-gems)
--template-file TEMPLATE                                  Full path to location of template to use
--use-template NAME                                       Try to use an existing template instead of importing disk
-r RUN_LIST | --run-list RUN_LIST                         Comma separated list of roles/recipes to apply
-j JSON_ATTRIBUTES | --json-attributes JSON_ATTRIBUTES    A JSON string to be added to the first run of chef-client
-x USERNAME | --ssh-user USERNAME                         The ssh username (default: root)
-P PASSWORD | --ssh-password PASSWORD                     The ssh password
-G GATEWAY | --ssh-gateway GATEWAY                        The ssh password
-i IDENTITY_FILE | --identity-file IDENTITY_FILE          The SSH identity file used for authentication
--no-host-key-verify                                      Disable host key verification
--vm-network network[,network..]                          Network where nic is attached to (default: 'VM Network')
-M mac[,mac..] | --mac-address mac[,mac..]                Mac address list
--skip-bootstrap                                          Skip bootstrap process (Deploy only mode)
--async                                                   Deploy the VMs asynchronously (Ignored unless combined with --batch)
--batch script.yml                                        Use a batch file to deploy multiple VMs
</pre>

### knife esx vm delete

Deletes an existing virtual machine in the currently configured ESX host by the virtual machine name. You can find the instance id by entering 'knife esx vm list'. Please note - this does not delete the associated node and client objects from the Chef server.

#### _Command line options_

    --force-delete YES|NO     Don't confirm the deletion if yes (default: no)

### knife esx vm list

Outputs a list of all virtual machines in the currently configured ESX host.  Please note - this shows all the virtual machines available in the ESX host, some of which may not be currently managed by the Chef server.

### knife esx template list

Outputs a list of all templates in the currently configured ESX host.

### knife esx template delete TEMPLATE_NAME

Deletes a template

#### _Command line options_

    --force-delete YES|NO     Don't confirm the deletion if yes (default: no)

## EXAMPLES

### Provision a new Ubuntu 11.10 VM using --vm-disk to import a local vmdk file.

    knife esx vm create --template-file ~/.chef/bootstrap/ubuntu11.10-gems.erb \
                        --vm-disk /path-to/ubuntu1110-x64-vmware-tools.vmdk \
                        --vm-name knife-esx-test-ubuntu \
                        --datastore datastore1
                        --esx-host my-test-host \
                        --esx-password secret

### Provision a new ubuntu 12.04 VM using --use-template to specify a template on the ESXi Host.

The command assumes that you have an ubuntu template located at /vmfs/volumes/datastore1/esx-gem/templates/ubuntu12.04-template-x64.vmdk.
In this example we also changed the location of the datastore from the default datastore1 to datastore2.

    knife esx vm create --esx-username root \
                        --vm-name ubuntu-12.04-vm \
                        --datastore datastore2 \
                        --esx-host server1 \
                        --esx-password secret \
                        --use-template ubuntu12.04-template-x64.vmdk

### Provision a new ubuntu 12.04 VM using --esx-templates-dir to specify a template directory on the ESXi host.

In this example, the template is located at /vmfs/volumes/datastore1/ubuntu12.04-template-x64/ubuntu12.04-template-x64.vmdk.

    knife esx vm create --esx-username root \
                        --vm-name ubuntu-12.04-vm \
                        --datastore datastore1 \
                        --esx-host server1 \
                        --use-template ubuntu12.04-template-x64.vmdk \
                        --esx-templates-dir /vmfs/volumes/datastore1/ubuntu12.04-template-x64
# LICENSE:

Author:: Sergio Rubio, Massimo Maino (<maintux@gmail.com>)
Copyright:: Copyright (c) 2011 Sergio Rubio, Massimo Maino
License:: Apache License, Version 2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# 0.3.1 - 2012/03/26

* Minor string change

# 0.3.0 - 2012/03/26

Added template commands

```
    knife esx template import --esx-password password \
                            --esx-host esx-test-host \
                            /path/to/template.vmdk

    knife esx template list --esx-password password \
                            --esx-host esx-test-host

    knife esx template delete --esx-password password \
                              --esx-host esx-test-host \
                              template.vmdk
```

Added --use-template argument to *vm create* command

```
    knife esx vm create --esx-password password \
                        --esx-host esx-test-host \
                        --use-template template.vmdk \
                        my-new-vm
```

## Create a new VM using templates

1.Import the template first with "knife esx template import"

```
    knife esx template import --esx-host esx-test-host \
                              --esx-password password \
                              /path/to/template.vmdk
```

2.Deploy using knife esx vm create

```
    knife esx vm create --esx-host esx-test-host \
                        --esx-password password \
                        --vm-name my-foo-vm \
                        --use-template template.vmdk
```

## Using templates with async batch deploys

1.Import the template first with "knife esx template import"

```
    knife esx template import --esx-host esx-test-host \
                              --esx-password password \
                              /path/to/template.vmdk
```

2.Deploy using knife esx vm create --async --batch
Sample batch config for "knife esx vm create":

```
    ---
    :tc01:
      'use-template': template.vmdk
      'extra-args': --no-host-key-verify
      'vm-memory': 128
      'esx-host': esx-test-host
      'esx-password': password
      'ssh-user': ubuntu
      'ssh-password': ubuntu
      'datastore': datastore1
    :tc02:
      'use-template': template.vmdk
      'extra-args': --no-host-key-verify
      'vm-memory': 128
      'esx-host': esx-test-host
      'esx-password': password
      'ssh-user': ubuntu
      'ssh-password': ubuntu
      'datastore': datastore1
```

# 0.2.1 - 2012/02/28

* Fixed namespaces

# 0.2 - 2012/02/28

* **Added --batch and --async options**

Inspired by spiceweasel from Matt Ray (https://github.com/mattray/spiceweasel), I've added a batch mode where a YAML file describes the VMs you want to bootstrap and where (you can deploy to multiple hypervisors).

    knife esx vm create --batch batch.yml

Sample batch.yml file:

    ---
    :test1:
      'extra-args': --no-host-key-verify
      'vm-memory': 128
      'esx-host': esx-server-1
      'esx-password': secret
      'ssh-user': ubuntu
      'ssh-password': ubuntu
      'vm-disk': /home/maintux/mnt/mirror/virtual_appliances/ubuntu1110-x64-vmware-tools.vmdk
      'datastore': datastore2
    :test2:
      'extra-args': --no-host-key-verify
      'vm-memory': 128
      'esx-host': esx-server-1
      'esx-password': secret
      'ssh-user': ubuntu
      'ssh-password': ubuntu
      'vm-disk': /home/maintux/mnt/mirror/virtual_appliances/ubuntu1110-x64-vmware-tools.vmdk
      'datastore': datastore2
    :test3:
      'extra-args': --no-host-key-verify
      'vm-memory': 256
      'esx-host': esx-server-1
      'esx-password': secret
      'ssh-user': ubuntu
      'ssh-password': ubuntu
      'vm-disk': /home/maintux/mnt/mirror/virtual_appliances/ubuntu1110-x64-vmware-tools.vmdk
      'datastore': datastore2

This will try to create three VMs (testvm1, testvm2 and testvm3) sequentially. VM definitions inside the batch file accept all the parameters that can be used with knife-esx.

If you want to bootstrap the VMs asynchronously, use the --async flag.

    knife esx vm create --batch batch.yml --async

When using batch mode, standard output and error is redirected to /tmp/knife_esx_vm_create_VMNAME.log, so if we use the deploy script from above, three log files will be created:

    /tmp/knife_esx_vm_create_test1.log
    /tmp/knife_esx_vm_create_test2.log
    /tmp/knife_esx_vm_create_test3.log

* **Added --skip-bootstrap flag**

If the flag is used the VM will be created but
  the bootstrap template/script won't be executed (it also means that Chef won't be installed).

* **Fixed bug preventing knife-esx to create a VM when the hypervisor has an empty root password.**

**KNOWN ISSUES**

* To use --batch without --skip-bootstrap, the ssh user (--ssh-user) needs to be able to sudo without asking for a password (i.e. adding something like 'ubuntu ALL=(ALL) NOPASSWD: ALL' to /etc/sudoers in the appliance template) otherwise the bootstraping process won't work if more than one VM is being deployed.

# 0.1.5 - 2012/02/25

* **Patch from @pperezrubio adding multiple networks and fixed MAC address support**

    knife esx vm create --vm-disk ubuntu-oneiric.vmdk \
                        --vm-name testvm --datastore datastore1 \
                        --esx-host 192.168.88.1 --ssh-user ubuntu \
                        --ssh-password ubuntu \
                        --vm-network "VLAN-Integration,VLAN-Test"

This will create a VM with two NICs, attaching them to the VLAN-Integration and VLAN-Test networks respectively.

Fixed MAC addresses can also be assigned to each NIC using the --mac-address parameter:

    knife esx vm create --vm-disk ubuntu-oneiric.vmdk \
                        --vm-name testvm --datastore datastore1 \
                        --esx-host 192.168.88.1 --ssh-user ubuntu \
                        --ssh-password ubuntu \
                        --vm-network "VLAN-Integration,VLAN-Test" \
                        --mac-address "00:01:02:03:04:05,00:01:02:03:04:06"

MAC address 00:01:02:03:04:05 will be assigned to VLAN-Integration NIC and 00:01:02:03:04:06 to the VLAN-Test NIC.

If a MAC address is omitted it will be dynamically generated:

knife esx vm create --vm-disk ubuntu-oneiric.vmdk \

lib/chef/knife/esx_vm_create.rb
#
# Author:: Sergio Rubio, Massimo Maino (<maintux@gmail.com>)
# Copyright:: Sergio Rubio, Massimo Maino (c) 2011
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'chef/knife/esx_base'
require 'open4'
require 'celluloid'
require 'singleton'
require 'yaml'

module KnifeESX
  class DeployScript

    attr_reader :job_count

    # Sample job
    #---
    #:test1:
    #  'vm-memory':
    #  'extra-args':
    #  'esx-host':
    #  'template-file':
    #  'vm-disk':
    #  'ssh-user':
    #  'ssh-password':
    #  'run-list':
    #  'network-interface':
    def initialize(batch_file)
      @batch_file = batch_file
      @jobs = []
      @job_count = 0
      (YAML.load_file batch_file).each do |i|
        @jobs << DeployJob.new(i)
        @job_count += 1
      end
    end

    def each_job(&block)
      @jobs.each do |j|
        yield j
      end
    end

  end

  class CLogger
    include Celluloid
    include Singleton

    def info(msg)
      puts "INFO: #{msg}"
    end

    def error(msg)
      $stderr.puts "ERROR: #{msg}"
    end
  end

  class DeployJob

    include Celluloid

    attr_reader :name

    def initialize(options)
      @name, @options = options
      validate
    end

    def validate
      if @name.nil? or @name.empty?
        raise Exception.new("Invalid job name")
      end
      if not @options['vm-disk'] or !File.exist?(@options['vm-disk'])
        raise Exception.new("Invalid VM disk for job #{@name}.")
      end
    end

    # returns [status, stdout, stderr]
    def run
      args = ""
      extra_args = ""
      @options.each do |k, v|
        if k == 'extra-args'
          extra_args << v
        else
          args << "--#{k} #{v} " unless k == 'extra-args'
        end
      end

      @out = ""
      @err = ""
      optstring = []
      @options.each do |k,v|
        optstring << "   - #{k}:".ljust(25) +  "#{v}\n" unless k =~ /password/
      end
      log_file = "/tmp/knife_esx_vm_create_#{@name.to_s.strip.chomp.gsub(/\s/,'_')}.log"
      KnifeESX::CLogger.instance.info! "Bootstrapping VM #{@name} \n#{optstring.join}"
      KnifeESX::CLogger.instance.info! "VM #{@name} bootstrap log: #{log_file}"
      @status = Open4.popen4("knife esx vm create --vm-name #{@name} #{args} #{extra_args} > #{log_file} 2>&1") do |pid, stdin, stdout, stderr|
        @out << stdout.read.strip
        @err << stderr.read.strip
      end
      if @status == 0
        KnifeESX::CLogger.instance.info! "[#{@name}] deployment finished OK"
      else
        KnifeESX::CLogger.instance.error! "[#{@name}] deployment FAILED"
        @err.each_line do |l|
          KnifeESX::CLogger.instance.error! "[#{@name}] #{l.chomp}"
        end
      end
      return @status, @out, @err
    end

  end

end

class Chef
  class Knife

    class EsxVmCreate < Knife

      include Knife::ESXBase

      deps do
        require 'readline'
        require 'esx'
        require 'chef/json_compat'
        require 'chef/knife/bootstrap'
        Chef::Knife::Bootstrap.load_deps
      end

      banner "knife esx vm create (options)"

      option :vm_disk,
        :long => "--vm-disk FILE",
        :description => "The path to the VMDK disk file"

      option :vm_name,
        :long => "--vm-name NAME",
        :description => "The Virtual Machine name"

      option :cpus,
        :long => "--vm-cpus CPUS",
        :description => "The Virtual Machine cpus (default: 1)"

      option :datastore,
        :long => "--datastore NAME",
        :default => 'datastore1',
        :description => "The Datastore to use for the VM files (default: datastore1)"

      option :guest_id,
        :long => "--guest-id NAME",
        :default => "otherGuest",
        :description => "The VM GuestID (default: otherGuest)"

      option :memory,
        :long => "--vm-memory MEM",
        :default => "512",
        :description => "The VM memory in MB (default: 512)"

      option :chef_node_name,
        :short => "-N NAME",
        :long => "--node-name NAME",
        :description => "The Chef node name for your new node"

      option :prerelease,
        :long => "--prerelease",
        :description => "Install the pre-release chef gems"

      option :bootstrap_version,
        :long => "--bootstrap-version VERSION",
        :description => "The version of Chef to install",
        :proc => Proc.new { |v| Chef::Config[:knife][:bootstrap_version] = v }

      option :distro,
        :short => "-d DISTRO",
        :long => "--distro DISTRO",
        :description => "Bootstrap a distro using a template; default is 'ubuntu10.04-gems'",
        :proc => Proc.new { |d| Chef::Config[:knife][:distro] = d },
        :default => "ubuntu10.04-gems"

      option :template_file,
        :long => "--template-file TEMPLATE",
        :description => "Full path to location of template to use",
        :proc => Proc.new { |t| Chef::Config[:knife][:template_file] = t },
        :default => false

      option :use_template,
        :long => "--use-template NAME",
        :description => "Try to use an existing template instead of importing disk",
        :default => nil

      option :run_list,
        :short => "-r RUN_LIST",
        :long => "--run-list RUN_LIST",
        :description => "Comma separated list of roles/recipes to apply",
        :proc => lambda { |o| o.split(/[\s,]+/) },
        :default => []

      option :json_attributes,
        :short => "-j JSON_ATTRIBUTES",
        :long => "--json-attributes JSON_ATTRIBUTES",
        :description => "A JSON string to be added to the first run of chef-client",
        :proc => lambda { |o| JSON.parse(o) },
        :default => {}

      option :ssh_user,
        :short => "-x USERNAME",
        :long => "--ssh-user USERNAME",
        :description => "The ssh username; default is 'root'",
        :default => "root"

      option :ssh_password,
        :short => "-P PASSWORD",
        :long => "--ssh-password PASSWORD",
        :description => "The ssh password"

      option :identity_file,
        :short => "-i IDENTITY_FILE",
        :long => "--identity-file IDENTITY_FILE",
        :description => "The SSH identity file used for authentication"

      option :no_host_key_verify,
        :long => "--no-host-key-verify",
        :description => "Disable host key verification",
        :boolean => true,
        :default => false,
        :proc => Proc.new { true }

      option :vm_network,
        :short => "-N network[,network..]",
        :long => "--vm-network",
        :description => "Network where nic is attached to",
        :default => 'VM Network'

      option :mac_address,
        :short => "-M mac[,mac..]",
        :long => "--mac-address",
        :description => "Mac address list",
        :default => nil

      option :skip_bootstrap,
        :long => "--skip-bootstrap",
        :description => "Skip bootstrap process (Deploy only mode)",
        :boolean => true,
        :default => false,
        :proc => Proc.new { true }

      option :async,
        :long => "--async",
        :description => "Deploy the VMs asynchronously (Ignored unless combined with --batch)",
        :boolean => true,
        :default => false,
        :proc => Proc.new { true }

      option :batch,
        :long => "--batch script.yml",
        :description => "Use a batch file to deploy multiple VMs",
        :default => nil

      def tcp_test_ssh(hostname)
        tcp_socket = TCPSocket.new(hostname, 22)
        readable = IO.select([tcp_socket], nil, nil, 5)
        if readable
          Chef::Log.debug("sshd accepting connections on #{hostname}, banner is #{tcp_socket.gets}")
          yield
          true
        else
          false
        end
      rescue Errno::ETIMEDOUT, Errno::EPERM
        false
      rescue Errno::ECONNREFUSED, Errno::EHOSTUNREACH, Errno::ENETUNREACH
        sleep 2
        false
      ensure
        tcp_socket && tcp_socket.close
      end

      def run
        $stdout.sync = true

        if config[:batch]
          KnifeESX::CLogger.instance.info "Running in batch mode. Extra arguments will be ignored."
          if not config[:async]
            counter = 0
            script = KnifeESX::DeployScript.new(config[:batch])
            script.each_job do |job|
              counter += 1
              status, stdout, stderr = job.run
              if status == 0
                KnifeESX::CLogger.instance.info 'Ok'
              else
                KnifeESX::CLogger.instance.error 'Failed'
                stderr.each_line do |l|
                  ui.error l
                end
              end
            end
          else
            KnifeESX::CLogger.instance.info! "Asynchronous boostrapping selected"
            KnifeESX::CLogger.instance.info! "Now do something productive while I finish my job ;)"
            script = KnifeESX::DeployScript.new(config[:batch])
            futures = []
            script.each_job do |job|
              futures << job.future(:run)
            end
            futures.each do |f|
              f.value
            end
          end
          return
        end

        if not config[:use_template]
          unless config[:vm_disk]
            ui.error("You have not provided a valid VMDK file. (--vm-disk)")
            exit 1
          end

          unless File.exist?(config[:vm_disk])
            ui.error("Invalid VMDK disk file (--vm-disk)")
            exit 1
          end
        end

        vm_name = config[:vm_name]
        if not vm_name
          ui.error("Invalid Virtual Machine name (--vm-name)")
          exit 1
        end


        datastore = config[:datastore]
        memory = config[:memory]
        cpus = config[:cpus]||1
        vm_disk = config[:vm_disk]
        guest_id =config[:guest_id]
        destination_path = "/vmfs/volumes/#{datastore}/#{vm_name}"

        connection.remote_command "mkdir #{destination_path}"
        ui.info "Creating VM #{vm_name}"

        if config[:use_template]
          ui.info "Using template #{config[:use_template]}"
          if connection.template_exist?(config[:use_template])
            puts "#{ui.color("Cloning template...",:magenta)}"
            connection.copy_from_template config[:use_template], destination_path + "/#{vm_name}.vmdk"
          else
            ui.error "Template #{config[:use_template]} not found"
            exit 1
          end
        else
          puts "#{ui.color("Importing VM disk... ", :magenta)}"
          connection.import_disk vm_disk, destination_path + "/#{vm_name}.vmdk"
        end
        vm = connection.create_vm :vm_name => vm_name,
                             :datastore => datastore,
                             :disk_file => "#{vm_name}/#{vm_name}.vmdk",
                             :memory => memory,
                             :cpus => cpus,
                             :guest_id => guest_id,
                             :nics => create_nics(config[:vm_network], config[:mac_address])
        vm.power_on

        puts "#{ui.color("VM Created", :cyan)}"
        puts "#{ui.color("VM Name", :cyan)}: #{vm.name}"
        puts "#{ui.color("VM Memory", :cyan)}: #{(vm.memory_size.to_f/1024/1024).round} MB"

        return if config[:skip_bootstrap]

        # wait for it to be ready to do stuff
        print "\n#{ui.color("Waiting server... ", :magenta)}"
        timeout = 100
        found = connection.virtual_machines.find { |v| v.name == vm.name }
        loop do
          if not vm.ip_address.nil? and not vm.ip_address.empty?
            puts "\n#{ui.color("VM IP Address: #{vm.ip_address}", :cyan)}"
            break
          end
          timeout -= 1
          if timeout == 0
            ui.error "Timeout trying to reach the VM. Does it have vmware-tools installed?"
            exit 1
          end
          sleep 1
          found = connection.virtual_machines.find { |v| v.name == vm.name }
        end

        print "\n#{ui.color("Waiting for sshd... ", :magenta)}"
        print(".") until tcp_test_ssh(vm.ip_address) { sleep @initial_sleep_delay ||= 10; puts(" done") }
        bootstrap_for_node(vm).run

        puts "\n"
        puts "#{ui.color("Name", :cyan)}: #{vm.name}"
        puts "#{ui.color("IP Address", :cyan)}: #{vm.ip_address}"
        puts "#{ui.color("Environment", :cyan)}: #{config[:environment] || '_default'}"
        puts "#{ui.color("Run List", :cyan)}: #{config[:run_list].join(', ')}"
        puts "#{ui.color("Done!", :green)}"
      end

      def bootstrap_for_node(vm)
        bootstrap = Chef::Knife::Bootstrap.new
        bootstrap.name_args = [vm.ip_address]
        bootstrap.config[:async] = config[:async]
        bootstrap.config[:run_list] = config[:run_list]
        bootstrap.config[:first_boot_attributes] = config[:json_attributes]
        bootstrap.config[:ssh_user] = config[:ssh_user]
        bootstrap.config[:identity_file] = config[:identity_file]
        bootstrap.config[:chef_node_name] = config[:chef_node_name] || vm.name
        bootstrap.config[:bootstrap_version] = locate_config_value(:bootstrap_version)
        bootstrap.config[:distro] = locate_config_value(:distro)
        # bootstrap will run as root...sudo (by default) also messes up Ohai on CentOS boxes
        bootstrap.config[:use_sudo] = true unless config[:ssh_user] == 'root'
        bootstrap.config[:template_file] = locate_config_value(:template_file)
        bootstrap.config[:environment] = config[:environment]
        bootstrap.config[:no_host_key_verify] = config[:no_host_key_verify]
        bootstrap.config[:ssh_password] = config[:ssh_password]
        bootstrap
      end

      def create_nics(networks, macs)
        net_arr = networks.split(/,/).map { |x| { :network => x } }
        if macs
          mac_arr = macs.split(/,/)
          net_arr.each_index { |x| net_arr[x][:mac_address] = mac_arr[x] if mac_arr[x] and !mac_arr[x].empty? }
        else
          net_arr
        end
      end
    end
  end
end
